crmprtd/process.py

```diff
@@ -111,21 +123,72 @@ def process(
         )
         raise ValueError("No network name given")
 
+    # Establish a database session
+    engine = create_engine(connection_string)
+    Session = sessionmaker(engine)
+    sesh = Session()
+
+    # For certain networks configured in `data/publication_lags.yaml`:
+    # If start_date is not specified, determine it from latest obs time in database.
+    # This query typically takes on the order of 90 s, but that cost is far lower than
+    # that of trying to insert - one by one - observations that are already present.
+    # A publication lag (days) is subtracted from the latest obs time to determine
+    # the start_date.
+
+    with resource_stream("crmprtd", "data/publication_lags.yaml") as pl_f:
+        publication_lags = yaml.safe_load(pl_f)
+
+    if network in publication_lags and start_date is None:
+        log.info(f"Querying database for latest obs time for network {network}...")
+        latest_obs_time = (
+            sesh.query(func.max(Obs.time))
+            .select_from(Obs)
+            .join(Variable)
+            .join(Network)
+            .filter(Network.name == "BCH")
+            .scalar()
+        )
+        log.info(f"Latest obs time for network {network}: {latest_obs_time}")
+        if latest_obs_time is not None:
+            # Obs.time is implicitly UTC but is unlocalized in the database.
+            # We have to make that explicit. Then subtract publication lag.
+            start_date = pytz.utc.localize(latest_obs_time) - timedelta(
+                days=(publication_lag or publication_lags[network])
+            )
+        else:
+            start_date = datetime.min
+
+    if network not in publication_lags and publication_lag is not None:
+        log.warning(
+            f"Publication lag (value {publication_lag}) is specified for "
+            f"network {network}, which does not use it."
+        )
+
+    # For all other cases, replace start_date == None with datetime.min.
+    if start_date is None:
+        start_date = datetime.min
+
+    # At present, this is impossible, but hey, it's cheap insurance.
+    if end_date is None:
+        start_date = datetime.max
+
+    log.info(
+        f"Final time filter parameters: "
+        f"start_date = {start_date}, end_date = {end_date}"
+    )
+
     # Get the normalizer for the specified network.
     download_stream = sys.stdin.buffer
     norm_mod = get_normalization_module(network)
 
-    # The normalizer returns a generator that yields `Row`s. Convert to a set of Rows.
     log.info("Normalize: start")
-
+    # The normalizer returns a generator that yields `Row`s. Convert to a set of Rows.
+    # It is probably better to use a dict for this to preserve order.
+    # See https://stackoverflow.com/a/9792680
```

Comment:
This SO comment is very out of date. I looked into this (should have noted it in the comment, b/c I can't find the evidence now), but contemporary Python sets *do* preserve insertion order.

Reply:
Hmmm... hold the presses on that... I may be incorrect here.